plc4j/protocols/ads/src/main/java/org/apache/plc4x/java/ads/protocol/Payload2TcpProtocol.java

```diff
@@ -49,21 +50,24 @@ protected void decode(ChannelHandlerContext channelHandlerContext, ByteBuf byteB
             return;
         }
         LOGGER.trace("(-->IN): {}, {}, {}", channelHandlerContext, byteBuf, out);
-        if (byteBuf.readableBytes() < AmsTcpHeader.Reserved.NUM_BYTES + TcpLength.NUM_BYTES) {
-            // wait till we can read the length header
-            return;
+
+        retainingBuf = Unpooled.wrappedBuffer(retainingBuf, byteBuf.retain());
+
+        while (retainingBuf.readableBytes() >= AmsTcpHeader.Reserved.NUM_BYTES + TcpLength.NUM_BYTES
+                && retainingBuf.readableBytes() >= retainingBuf.getUnsignedIntLE(retainingBuf.readerIndex()
+                        + AmsTcpHeader.Reserved.NUM_BYTES) + AmsTcpHeader.Reserved.NUM_BYTES + TcpLength.NUM_BYTES) {
+            // Reserved
+            retainingBuf.skipBytes(AmsTcpHeader.Reserved.NUM_BYTES);
+            TcpLength packetLength = TcpLength.of(retainingBuf);
+            AmsTcpHeader amsTcpHeader = AmsTcpHeader.of(packetLength);
+            LOGGER.debug("AMS TCP Header {}", amsTcpHeader);
+
+            out.add(retainingBuf.readBytes((int) packetLength.getAsLong()));
         }
-        if (byteBuf.readableBytes() < byteBuf.getUnsignedIntLE(AmsTcpHeader.Reserved.NUM_BYTES)) {
-            // wait till we have a complete ADS packet
-            return;
+        if (retainingBuf.readableBytes() == 0) {
```

Comment:
Could this lead to a memory leak? We should also release the byteBuf, right? Currently retain is called (L54) but never release.

Reply:
The documentation: https://netty.io/4.0/api/io/netty/buffer/Unpooled.html#wrappedBuffer-io.netty.buffer.ByteBuf...-
States that the reference count ownership is transfered to this method. So when releasing (L68) the byteBuf should be released.